packages/data-viz/src/core/HeatmapChart/index.tsx

```diff
@@ -18,101 +19,176 @@ export interface HeatmapChartProps
-    throw Error("Chart must have width and height > 0");
-  }
+const HeatmapChart = forwardRef(
+  (
+    props: HeatmapChartProps,
+    ref: ForwardedRef<HTMLDivElement>
+    // eslint-disable-next-line sonarjs/cognitive-complexity
+  ): JSX.Element => {
+    const {
+      axisPointer,
+      width,
+      height,
+      echartsRendererMode = "svg",
+      camera,
+      onEvents,
+      xAxisData,
+      yAxisData,
+      data,
+      encode,
+      emphasis,
+      itemStyle,
+      symbol,
+      symbolSize = 5,
+      grid,
+      options,
+      ...rest
+    } = props;
+
+    // Validate width and height
+    if (!width || !height) {
+      throw Error("Chart must have width and height > 0");
+    }
 
-  const [isChartInitialized, setIsChartInitialized] = useState(false);
+    // State for chart initialization
+    const [isChartInitialized, setIsChartInitialized] = useState(false);
+
+    // State for ECharts instance
+    const [chart, setChart] = useState<ECharts | null>(null);
+
+    // Ref for the chart container
+    const innerRef = useRef<HTMLDivElement | null>(null);
+
+    // Function to initialize the chart
+    const initChart = useCallback(() => {
+      const { current } = innerRef;
+
+      if (
+        !current ||
+        isChartInitialized ||
+        // (thuang): echart's `init()` will throw error if the container has 0 width or height
+        current?.getAttribute("height") === "0" ||
+        current?.getAttribute("width") === "0"
+      ) {
+        return;
+      }
+
+      // Initialize ECharts instance
+      const rawChart = init(current, EMPTY_OBJECT, {
+        renderer: echartsRendererMode,
+        useDirtyRect: true,
+      });
 
-  const [chart, setChart] = useState<ECharts | null>(null);
-  const innerRef = useRef<HTMLDivElement | null>(null);
+      setIsChartInitialized(true);
+
+      // Bind events if provided
+      if (onEvents) {
+        bindEvents(rawChart, onEvents);
+      }
+
+      setChart(rawChart);
+
+      // Cleanup function
+      return () => {
+        chart?.dispose();
+
+        // Unbind events if provided
+        if (onEvents) {
+          for (const eventName in onEvents) {
+            if (
+              Object.prototype.hasOwnProperty.call(onEvents, eventName) &&
+              typeof eventName === "string" &&
+              typeof onEvents[eventName] === "function"
+            ) {
+              rawChart.off(eventName, onEvents[eventName]);
+            }
+          }
+        }
+      };
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [chart, echartsRendererMode, isChartInitialized, onEvents]);
+
+    // Initialize charts on component mount
+    useEffect(() => {
+      initChart();
+    }, [initChart]);
+
+    // Reinitialize chart if renderer mode changes
+    useEffect(() => {
+      chart?.dispose();
+      setIsChartInitialized(false);
+      setChart(null);
+      initChart();
+      // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
same here 😁

Reply:
We don't want to include the `chart` dependency in the array because doing so would lead to an infinite loop of initializing and disposing the chart. We specifically want to reinitialize the chart only when the renderer mode changes. However, since we use the `chart` state to dispose of the chart, we disable the `eslint` rule for `exhaustive-deps` to prevent warnings.

Reply:
Ah got it! Thanks for the context, @masoudmanson 💡 

I'm not sure if this helps solve the problem, but I have a PR to use a chart useState AND a useRef, so some hooks can opt out of using the chart state, which causes infinite loop

https://github.com/chanzuckerberg/sci-components/pull/749

What do you think? Is this more confusing than its benefits lol? 

Thank you!

Reply:
Oh, awesome, let me take a look at it. 😍